jupyterlab/extensions/manager.py

```diff
@@ -161,7 +177,117 @@ class ExtensionsCache:
     last_page: int = 1
 
 
-class ExtensionManager(LoggingConfigurable):
+class PluginManager(LoggingConfigurable):
+    """Plugin manager enables or disables plugins unless locked.
+
+    It can also disable/enable all plugins in an extension.
+
+    Args:
+        app_options: Application options
+        ext_options: Plugin manager (subset of extension manager) options
+        parent: Configurable parent
+
+    Attributes:
+        app_options: Application options
+        options: Plugin manager options
+    """
+
+    def __init__(
+        self,
+        app_options: Optional[dict] = None,
+        ext_options: Optional[dict] = None,
+        parent: Optional[Configurable] = None,
+    ) -> None:
+        super().__init__(parent=parent)
+        self.app_options = _ensure_options(app_options)
+        plugin_options_field = {f.name for f in fields(PluginManagerOptions)}
+        plugin_options = {
+            option: value
+            for option, value in (ext_options or {}).items()
+            if option in plugin_options_field
+        }
+        self.options = PluginManagerOptions(**plugin_options)
+
+    async def plugin_locks(self) -> dict:
+        """Get information about locks on plugin enabling/disabling"""
+        return {
+            "lockRules": list(self.options.lock_rules),
+            "allLocked": self.options.lock_all,
+        }
+
+    def _find_locked(self, plugins_or_extensions: List[str]) -> Set[str]:
+        """Find a subset of plugins (or extensions) which are locked"""
+        if self.options.lock_all:
+            return set(plugins_or_extensions)
+        locked_subset = set()
+        extensions_with_locked_plugins = {
+            plugin.split(':')[0] for plugin in self.options.lock_rules
+        }
+        for plugin in plugins_or_extensions:
+            if ':' in plugin:
+                # check directly if this is a plugin identifier (has colon)
```

Comment:
Curious about what would happen with third-party extensions that don't respect using the `:` delimiter? Would all the single plugins be seen as extensions?

Reply:
At the CLI level yes. At the frontend level this is an existing issue in the extension manager too (disabling extension only works if it uses nice plugin names), however, this PR adds `availablePlugins` which allows to map between individual plugins and extensions so now have a way of fixing it in extension manager UI.